client/src/components/logViewer/logParser.ts

```diff
@@ -0,0 +1,438 @@
+    ["codeLogLine", codeLogLine, "replace"],
+    ["logLines", [], "replace"],
+    ["type", "error", "replace"],
+  ]);
+}
+
+function removeHandledProblemIndex(problemMetadata: ProblemMetadata) {
+  if (problemMetadata.problemIndicsForSameLocation.length > 0) {
+    const handledProblemIndex = getProblemIndexFromProblemBeginningLogLine(
+      problemMetadata.logLines[0],
+    );
+
+    const newIndices: string[] = [];
+    problemMetadata.problemIndicsForSameLocation.forEach((problemIndex) => {
+      if (problemIndex !== handledProblemIndex) {
+        newIndices.push(problemIndex);
+      }
+    });
+
+    updateProblemMetadata(problemMetadata, [
+      ["problemIndicsForSameLocation", newIndices, "replace"],
+    ]);
+  }
+}
+
+function calculateLocationOffset(
+  codeLogLine: string,
+  firstCodeLine: string,
+): { columnOffset: number; lineOffset: number } {
+  const codeInfo = extractCodeInfoFromCodeLogLine(codeLogLine);
+
+  // there may be a log kept when finishing running selected code,
+  // the kept log will be sent out in following running code.
+  // that log line number should not be used for calculating line offset.
+  if (codeInfo === null || firstCodeLine !== codeInfo.code.trim()) {
+    return { lineOffset: -1, columnOffset: -1 };
+  }
+
+  const lineOffset = codeInfo.lineNumber;
+  const regexToMatchFirstNonSpaces = /\w/;
+  const firstCharIndexInCode =
+    firstCodeLine.match(regexToMatchFirstNonSpaces)?.index ?? -1;
+  const firstCharIndexInLog =
+    codeInfo.code.match(regexToMatchFirstNonSpaces)?.index ?? -1;
+  const lineNumberLength = codeInfo.lineNumber.toString().length;
+  const columnOffset =
+    firstCharIndexInLog - firstCharIndexInCode + lineNumberLength;
+
+  return { columnOffset, lineOffset };
+}
+
+function getFirstCharacterColumn(line: string): number {
+  const regExp = /[^\s]/;
+  if (line.trim() === "") {
+    return -1;
+  }
+
+  const match = line.match(regExp);
+  return match === null || match.index === undefined ? -1 : match.index;
+}
+
+function handleProblemMetadata(
+  problems: Problem[],
+  problemMetadata: ProblemMetadata,
+  codeLogLine: string,
+  offset: LocationOffset,
+) {
+  if (hasNotCachedProblem(problemMetadata)) {
+    return;
+  }
+
+  handleProblemLocation(problemMetadata, codeLogLine, offset);
+
+  addProblem(problems, problemMetadata);
+
+  resetProblemMetadata(problemMetadata);
+}
+
+function hasNotCachedProblem(problemMetadata: ProblemMetadata): boolean {
+  return problemMetadata.logLines.length === 0;
+}
+
+function addProblem(problems: Problem[], problemMetadata: ProblemMetadata) {
+  problems.push(constructProblem(problemMetadata));
+}
+
+function handleProblemLocation(
+  problemMetadata: ProblemMetadata,
+  codeLogLine: string,
+  offset: LocationOffset,
+) {
+  if (!isValidProblemLocation(problemMetadata.location)) {
+    problemMetadata.location = generateLocationFromCodeLogLine(
+      codeLogLine,
+      offset,
+    );
+  }
+}
+
+//"      ---"
```

Comment:
Did you mean to leave this in? Oh, I see. Would you mind prefixing with something like `// Example: "      ---"`?